cmd/dryrun/dryrun.go

```diff
@@ -0,0 +1,478 @@
+			} else if err != nil {
+				msg := fmt.Sprintf("Failed to setup %s. %v", comp.Name(), err)
+				log.Warningf(msg)
+			}
+		} else {
+			if err := comp.Delete(cluster); err != nil {
+				msg := fmt.Sprintf("Failed to cleanup %v. %v", comp.Name(), err)
+				log.Warningf(msg)
+			}
+		}
+	}
+
+	controller := storagecluster.Controller{Driver: dryRunDriver}
+	controller.SetKubernetesClient(dryRunClient)
+	objs, err := getAllObjects(controller, cluster)
+	if err != nil {
+		log.WithError(err).Error("failed to get all objects")
+	}
+
+	outputFile := c.String(flagOutputFile)
+	if outputFile == "" {
+		outputFile = defaultOutputFile
+	}
+
+	var sb strings.Builder
+	for _, obj := range objs {
+		log.Infof("Operator will deploy %s %s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetName())
+		bytes, err := yaml.Marshal(obj)
+		if err != nil {
+			return err
+		}
+
+		sb.WriteString(string(bytes))
+		sb.WriteString("---\n")
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+	if err != nil {
+		log.WithError(err).Errorf("failed to open file %s", outputFile)
+		return err
+	}
+	defer f.Close()
+	_, err = f.WriteString(sb.String())
+	if err != nil {
+		log.WithError(err).Errorf("failed to write file %s", outputFile)
+		return err
+	}
+
+	log.Infof("wrote %d objects to file %s after dry run", len(objs), outputFile)
+	return nil
+}
+
+func getStorageCluster(c *cli.Context) (*corev1.StorageCluster, error) {
+	storageClusterFile := c.String(flagStorageCluser)
+	if storageClusterFile != "" {
+		objs, err := inttestutil.ParseSpecsWithFullPath(storageClusterFile)
+		if err != nil {
+			return nil, err
+		}
+		if len(objs) != 1 {
+			return nil, fmt.Errorf("require 1 object in the config file, found objects: %+v", objs)
+		}
+		cluster, ok := objs[0].(*corev1.StorageCluster)
+		if !ok {
+			return nil, fmt.Errorf("object is not storagecluster, %+v", objs[0])
+		}
+		return cluster, nil
+	}
+
+	k8sClient, err := getK8sClient(c)
+	if err != nil {
+		return nil, err
+	}
+
+	nsList := &v1.NamespaceList{}
+	err = k8sClient.List(context.TODO(), nsList)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, namespace := range nsList.Items {
+		clusterList := &corev1.StorageClusterList{}
+		err = k8sClient.List(context.TODO(), clusterList, &client.ListOptions{
+			Namespace: namespace.Name,
+		})
+		if err != nil {
+			log.WithError(err).Error("failed to list StorageCluster")
+		}
+		if len(clusterList.Items) > 0 {
+			return &clusterList.Items[0], nil
+		}
+	}
+
+	return nil, fmt.Errorf("could not find storagecluster object from k8s")
+}
+
+func getK8sClient(c *cli.Context) (client.Client, error) {
+	var err error
+	var config *rest.Config
+	kubeconfig := c.String(flagKubeConfig)
+	if kubeconfig == "" {
+		kubeconfig = os.Getenv("KUBECONFIG")
+	}
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+		defaultOutputFile = "/tmp/" + defaultOutputFile
```

Comment:
is there anything to do with this defaultOutputFile here?

Reply:
the reason is when running inside of container, it would get "permission denied" error writing to root, hence writing it to /tmp. I will add a comment.